src/matrix/storage/idb/Store.ts

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2020 Bruno Windels <bruno@windels.cloud>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {QueryTarget, IDBQuery} from "./QueryTarget";
+import {IDBRequestAttemptError} from "./error";
+import {reqAsPromise} from "./utils";
+import {Transaction} from "./Transaction";
+
+const LOG_REQUESTS = false;
+
+function logRequest(method: string, params: any[], source: any): void {
+    const storeName = source?.name;
+    const databaseName = source?.transaction?.db?.name;
+    console.info(`${databaseName}.${storeName}.${method}(${params.map(p => JSON.stringify(p)).join(", ")})`);
+}
+
+class QueryTargetWrapper<T> {
+    private _qt: IDBIndex | IDBObjectStore;
+
+    constructor(qt: IDBIndex | IDBObjectStore) {
+        this._qt = qt;
+    }
+
+    get keyPath(): string | string[] {
+        if (this._qt["objectStore"]) {
```

Comment:
So had a look, and the idiomatic way to narrow the type seems to be [`in` narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) here, which preserves type-safety without casts. Also see [this playground example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQwM4GUpwE4FMAUAjlAFyICSAIgELlgAmuAHogD4U0DyARgFa7RMOXAEoyVaj36CseRAG8AUIhWIYwRPgBEcPgKhC8WtUmIiFy1VbxQQ2U1AB0u6Qdm4A3JZUBfb4hs7By8-RVBIWAQUDHcAJiJSDloGZjYkqX1DUXEuPRlhCyt1TWIAbR08t2EtAF1zJSsrYmdKrK8rP2tcW3tEYhDFIA).